README.md
# mast-session

A Clojure library designed to ... well, that part is up to you.

## Usage

This will interact with firebase jwt at one edge and will be largely difficult to test in that regard. Tokens expire and must be signed with a private key.

These keys can be generated using cljsjs/firebase "4.2.0" and the config info from the Firebase Setup https://console.firebase.google.com/project/breezeehr.com:breeze-ehr/authentication/users  and clicking on "Web Setup".

```clojure

(defn register-firebase [config]
  (.initializeApp js/firebase config))

;; hook up app onces to firebase
(defonce registered-with-firebase
  (do (register-firebase firebase-config)
      true))

(defn authenticate [username password]
  (let [c (async/chan)]
    (-> (.auth js/firebase)
      (.setPersistence persistence)
      (.then (fn []
               (.signInWithEmailAndPassword (.auth js/firebase)  username password)))
      (.then
        (fn [result]
          (js/console.log "worked:" result "worked")
          (go (async/>! c result)))
        (fn [err]
          (let [e (which-error err)]
            (js/console.log e)
            (go (async/>! c e))))))
    c))

(let [chan         (authenticate "dsutton@breezeehr.com" "easypassword")
      forceRefresh false]

    (go
      (let [user (async/<! chan identity) ]
        (-> (.getIdToken user forceRefresh)
          (.then
            (fn [token]
              (js/console.log token))
            (fn [err] err))))))

```

This will print a token which can be decoded from `mast-session.core/decode-object` into something like the following:

```clojure

{:firebase
 {:identities {:email ["dsutton@breezeehr.com"]},
  :sign_in_provider "password"},
 :email "dsutton@breezeehr.com",
 :aud "breezeehr.com:breeze-ehr",
 :sub "Spgusa58rXSxAcfL0dab5yBP3Iz2",
 :iss "https://securetoken.google.com/breezeehr.com:breeze-ehr",
 :exp 1503090344,
 :email_verified false,
 :auth_time 1503086744,
 :user_id "Spgusa58rXSxAcfL0dab5yBP3Iz2",
 :iat 1503086744}

```
## License

Copyright © 2017 BreezeEHR


CHANGELOG.md
# Change Log
All notable changes to this project will be documented in this file. This change log follows the conventions of [keepachangelog.com](http://keepachangelog.com/).

## [Unreleased]
### Changed
- Add a new arity to `make-widget-async` to provide a different widget shape.

## [0.1.1] - 2017-08-18
### Changed
- Documentation on how to make the widgets.

### Removed
- `make-widget-sync` - we're all async, all the time.

### Fixed
- Fixed widget maker to keep working when daylight savings switches over.

## 0.1.0 - 2017-08-18
### Added
- Files from the new template.
- Widget maker public API - `make-widget-sync`.

[Unreleased]: https://github.com/your-name/mast-session/compare/0.1.1...HEAD
[0.1.1]: https://github.com/your-name/mast-session/compare/0.1.0...0.1.1


test/mast_session/core_test.clj
(ns mast-session.core-test
  (:require [clojure.test :refer :all]
            [mast-session.core :refer :all]))

(deftest a-test
  (testing "FIXME, I fail."
    (is (= 0 1))))


test/mast_session/token_test.clj
(ns mast-session.token-test
  (:require [clojure.test :refer :all]
            [mast-session.token :refer :all]
            [mast-session.certs :refer [google-certs]]
            [buddy.core.keys :as ks]))

(def tkn "eyJhbGciOiJSUzI1NiIsImtpZCI6IjQ4YjBmODA4MmE3YmUyZGIxNTAyNjhhYzZiNWNjODJhMjliZWRkMDgifQ.eyJpc3MiOiJodHRwczovL3NlY3VyZXRva2VuLmdvb2dsZS5jb20vYnJlZXplZWhyLmNvbTpicmVlemUtZWhyIiwiYXVkIjoiYnJlZXplZWhyLmNvbTpicmVlemUtZWhyIiwiYXV0aF90aW1lIjoxNTAzMDg2NzQ0LCJ1c2VyX2lkIjoiU3BndXNhNThyWFN4QWNmTDBkYWI1eUJQM0l6MiIsInN1YiI6IlNwZ3VzYTU4clhTeEFjZkwwZGFiNXlCUDNJejIiLCJpYXQiOjE1MDMwODY3NDQsImV4cCI6MTUwMzA5MDM0NCwiZW1haWwiOiJkc3V0dG9uQGJyZWV6ZWVoci5jb20iLCJlbWFpbF92ZXJpZmllZCI6ZmFsc2UsImZpcmViYXNlIjp7ImlkZW50aXRpZXMiOnsiZW1haWwiOlsiZHN1dHRvbkBicmVlemVlaHIuY29tIl19LCJzaWduX2luX3Byb3ZpZGVyIjoicGFzc3dvcmQifX0.RGNOamxM_l_OLw2xvSCyjz1ra4dBFLSYj-fxcEFL_KBHn-Ur46JQi763lY28wPxtxvlxGg8PGkV4eoEF7qOKOGnweZdOt2YEELmDN7QWWAtz04_T41DRCm0nnPbhiO3VuCnDpWOiVD8NFS0fUj-LXlyRHxCegmPqadibR_ItNUeUNbPogGO8TeIHvyr9hWvMQHQbYYXUcNO0PFJZfhOq-UfdiLPYVtP6lLMDQOv30kiMOQjJGq_L6vO654eW5U1quG6cKupCmJoDtRMgVyP_F5_3QvMCKD3Enxn7jPWXGERJLApv5JZrBZEaPFz51R1aqyRFyfJafaZnb6HylsU3Tg")

(def cert "-----BEGIN CERTIFICATE-----\nMIIDHDCCAgSgAwIBAgIINQZNrvqfzWUwDQYJKoZIhvcNAQEFBQAwMTEvMC0GA1UE\nAxMmc2VjdXJldG9rZW4uc3lzdGVtLmdzZXJ2aWNlYWNjb3VudC5jb20wHhcNMTcw\nODIxMDA0NTI2WhcNMTcwODI0MDExNTI2WjAxMS8wLQYDVQQDEyZzZWN1cmV0b2tl\nbi5zeXN0ZW0uZ3NlcnZpY2VhY2NvdW50LmNvbTCCASIwDQYJKoZIhvcNAQEBBQAD\nggEPADCCAQoCggEBALiOCSl8rcCPdcFY6Zdd1Oicgp+fAd6WmvgDjW06umtpPKZ3\nLcAtaN70GzEkN9J5IugPjBKQlKxEyw6SlA8ByTqlXZ1rOE+0NL1GttdkjW3E+3Aw\nWDJLNLggiO+5+6GyC3/HMxHzJuqOJlsaddoxxtVbG01sbjukO6YDcekvmlwUkl0R\nzqXDVvC1QEUQSGzbge4spvw6ItF6zAxOqTJIDyaxIeCG15Ns5AAIgrAaQFC9tql7\neOKJccuX0ReO/D4cMG2tt1fGTr9UmII/6oNm5Gko1gj21Zu1JvsIWiyDtYIcBXj7\nwKnr0BAO9/M+VZK8pDir4vmXXcWeUYrZEa4U4b0CAwEAAaM4MDYwDAYDVR0TAQH/\nBAIwADAOBgNVHQ8BAf8EBAMCB4AwFgYDVR0lAQH/BAwwCgYIKwYBBQUHAwIwDQYJ\nKoZIhvcNAQEFBQADggEBAJwbYp7iqYWFv4SCAK1YVNsCt7M2Y8U11U9kPkJe3FHo\nb+FyqMi14SurHwSnGWaoEcWUzaQ36K2f+RxAKuoOLFK9O6HtcLl0LP7ge84VWZL1\nJL4Mdpow6GcLly7AD93zYFLP7pHjB5EiCzGTHaCA2sCD24xcEe8Fxf3/2K1EOm91\n2qTImG4GAzp4bech4/uJH0YAsTWFb+MGnAVR+lVDgoz/DyYvvwwZTIw2o3W5I4R9\nHHcp+U99ZxfMGf9ttpn+0DihhkhqlMlVvAw4h8fcxdxJ3m9X9AqBp8Y5zKyzaCyJ\nEfN5CX5CpVjp3kNh5k6BVht2sMqgW3pV09LR1T8vOug=\n-----END CERTIFICATE-----\n")

(def token-contents {:firebase
                     {:identities {:email ["dsutton@breezeehr.com"]},
                      :sign_in_provider "password"},
                     :email "dsutton@breezeehr.com",
                     :aud "breezeehr.com:breeze-ehr",
                     :sub "Spgusa58rXSxAcfL0dab5yBP3Iz2",
                     :iss "https://securetoken.google.com/breezeehr.com:breeze-ehr",
                     :exp 1503090344,
                     :email_verified false,
                     :auth_time 1503086744,
                     :user_id "Spgusa58rXSxAcfL0dab5yBP3Iz2",
                     :iat 1503086744})

(def cert (-> google-certs
            vals
            last
            ks/str->public-key))

(def testing-options {:leeway 1000000000000}) ;; allow for expired certs

(deftest decode-object-test
  (testing "can read firebase tokens"
    (is (= "dsutton@breezeehr.com"
          (:email (decode-google-payload tkn cert
                    (merge default-token-options testing-options)))))))

(deftest firebase->user-tests
  (testing "makes all datetimes good to go"
    (let [time-vals #{:expires :auth-time :issued-at}]
      (is (= #{org.joda.time.DateTime})
        (->> (select-keys (firebase->user token-contents)
               time-vals)
          (map class)
          set)))))


test/mast_session/session_test.clj
(ns mast-session.session-test
  (:require [mast-session.session :refer :all]
            [clojure.test :refer [deftest is testing]]
            [clj-time.core :as t]))

(defn user-auth-time
  ([]
   (user-auth-time 1))
  ([minutes-ago]
   (user-auth-time minutes-ago {:user-id 1}))
  ([minutes-ago user]
   (merge user
     {:auth-time (t/plus (t/now) (t/minutes (- minutes-ago)))})))

(deftest can-add-user?-tests
  (testing "if recent token then yes"
    (is (can-add-user? (user-auth-time 2)))
    (is (not (can-add-user? (user-auth-time 4))))))

(deftest has-session?-tests
  (testing "when no active session"
    (let [session (atom {})
          user    (user-auth-time)]
      (is (not (has-session? user session))))
    (let [user-id 1
          user    (user-auth-time 5
                    {:user-id     user-id
                     :last-access (t/minus (t/now)
                                    recent-threshold
                                    (t/minutes 1))})
          session (atom {user-id user})]
      (is (not (has-session? user session)))))

  (testing "user present and within the time threshold"
    (let [user-id 1
          user    (user-auth-time 5
                    {:user-id     user-id
                     :last-access (t/plus
                                    (t/minus (t/now)
                                      recent-threshold)
                                    (t/minutes 1))})
          session (atom {user-id user})]
      (is (= true (has-session? user session)))))
  (testing "user has session but ticket is old"
    (let [user-id   1
          last-active (t/plus (t/now) (t/minutes -1))
          issued-at   (t/plus (t/now) (t/minutes -200))
          user        (user-auth-time 5
                        {:user-id     user-id
                         :issued-at  issued-at})
          session     (atom {user-id {:user-id     user-id
                                      :last-access last-active}})]
      (is (not (has-session? user session))))))


src/mast_session/certs.clj
(ns mast-session.certs
  (:require [clj-http.client :as client]
            [cheshire.core :as json]))

;;; These certs are from
;;; https://www.googleapis.com/robot/v1/metadata/x509/securetoken@system.gserviceaccount.com

;; Finally, ensure that the ID token was signed by the private key
;; corresponding to the token's kid claim. Grab the public key from
;; https://www.googleapis.com/robot/v1/metadata/x509/securetoken@system.gserviceaccount.com
;; and use a JWT library to verify the signature. Use the value of
;; max-age in the Cache-Control header of the response from that
;; endpoint to know when to refresh the public keys.

;; https://firebase.google.com/docs/auth/admin/verify-id-tokens


(def google-certs
  {"827177af78ca96b9460c0748a0c728352053f136" "-----BEGIN CERTIFICATE-----\nMIIDHDCCAgSgAwIBAgIINQZNrvqfzWUwDQYJKoZIhvcNAQEFBQAwMTEvMC0GA1UE\nAxMmc2VjdXJldG9rZW4uc3lzdGVtLmdzZXJ2aWNlYWNjb3VudC5jb20wHhcNMTcw\nODIxMDA0NTI2WhcNMTcwODI0MDExNTI2WjAxMS8wLQYDVQQDEyZzZWN1cmV0b2tl\nbi5zeXN0ZW0uZ3NlcnZpY2VhY2NvdW50LmNvbTCCASIwDQYJKoZIhvcNAQEBBQAD\nggEPADCCAQoCggEBALiOCSl8rcCPdcFY6Zdd1Oicgp+fAd6WmvgDjW06umtpPKZ3\nLcAtaN70GzEkN9J5IugPjBKQlKxEyw6SlA8ByTqlXZ1rOE+0NL1GttdkjW3E+3Aw\nWDJLNLggiO+5+6GyC3/HMxHzJuqOJlsaddoxxtVbG01sbjukO6YDcekvmlwUkl0R\nzqXDVvC1QEUQSGzbge4spvw6ItF6zAxOqTJIDyaxIeCG15Ns5AAIgrAaQFC9tql7\neOKJccuX0ReO/D4cMG2tt1fGTr9UmII/6oNm5Gko1gj21Zu1JvsIWiyDtYIcBXj7\nwKnr0BAO9/M+VZK8pDir4vmXXcWeUYrZEa4U4b0CAwEAAaM4MDYwDAYDVR0TAQH/\nBAIwADAOBgNVHQ8BAf8EBAMCB4AwFgYDVR0lAQH/BAwwCgYIKwYBBQUHAwIwDQYJ\nKoZIhvcNAQEFBQADggEBAJwbYp7iqYWFv4SCAK1YVNsCt7M2Y8U11U9kPkJe3FHo\nb+FyqMi14SurHwSnGWaoEcWUzaQ36K2f+RxAKuoOLFK9O6HtcLl0LP7ge84VWZL1\nJL4Mdpow6GcLly7AD93zYFLP7pHjB5EiCzGTHaCA2sCD24xcEe8Fxf3/2K1EOm91\n2qTImG4GAzp4bech4/uJH0YAsTWFb+MGnAVR+lVDgoz/DyYvvwwZTIw2o3W5I4R9\nHHcp+U99ZxfMGf9ttpn+0DihhkhqlMlVvAw4h8fcxdxJ3m9X9AqBp8Y5zKyzaCyJ\nEfN5CX5CpVjp3kNh5k6BVht2sMqgW3pV09LR1T8vOug=\n-----END CERTIFICATE-----\n",
   "08e1305a833e370ca65cc817f2fc6a643a92024e" "-----BEGIN CERTIFICATE-----\nMIIDHDCCAgSgAwIBAgIIHl6tlmXc1SswDQYJKoZIhvcNAQEFBQAwMTEvMC0GA1UE\nAxMmc2VjdXJldG9rZW4uc3lzdGVtLmdzZXJ2aWNlYWNjb3VudC5jb20wHhcNMTcw\nODE5MDA0NTI2WhcNMTcwODIyMDExNTI2WjAxMS8wLQYDVQQDEyZzZWN1cmV0b2tl\nbi5zeXN0ZW0uZ3NlcnZpY2VhY2NvdW50LmNvbTCCASIwDQYJKoZIhvcNAQEBBQAD\nggEPADCCAQoCggEBAKKGigwbHVJYm1d82qVlPycCJ7eza+FCkeQIH4rMF7/45YCq\nvbC9YYoFBeleySWEteb45rlOzWVPLo2MyYuodmkVcP63hD3AzAErMMoHgd8yHvhe\nzYsfiD4wRjCezuJ9uod4aPuJ+GFvwxG9FWqh1rr/Im/Fp7MIQMlwBBIRhlB8m4Ob\nX8S8KLdNZFKLuYJoGBVaypsddvcZMshxM3eaB82N1kNsMGcljfDgJpepm6ECm6HM\nmdYmG5f6rUnIsAH8jnuo6H34xBsn/S5kM6td7Jcv2+8lnYasfSJ88QiFJTzLmdMQ\naKUkHLe6fesD5Pl0ncalqsyFo+Dr6Ot7c8pVh8cCAwEAAaM4MDYwDAYDVR0TAQH/\nBAIwADAOBgNVHQ8BAf8EBAMCB4AwFgYDVR0lAQH/BAwwCgYIKwYBBQUHAwIwDQYJ\nKoZIhvcNAQEFBQADggEBAIvLWN1ubiQ/3vts+3JzDI3PrGkvM2CSWp+JahdZ9Wqq\nFV4Z/lN9n9iuwDh5RJEDQQzuWVumRsj9PNAfolAt8F6uX7C0tDqr44SgQW7uiT3h\nbQm7Hk7onujZz4Ba/tFNKXFjJ+zE1Aa1FwwhqkIo9VcLbExc+XzWo03HWucsotZs\nIDWMoGH1GAWdzfD+m1Tj4OATa/shitqjYrfBeGj4Y+SbUCCP/uO9/AKlvtHtBii/\nUZCOTpWmIEupxGqjUJysVPe32IDJSikexLd9vsvRXJ6fneyCG6PVEpLVcaifYtcA\nrv4PvSwl6UY7Kg6clRU3DaZldETYZzS3PE++RF4f4b4=\n-----END CERTIFICATE-----\n",
   "c947c408c8dd053f7e13117c4e00f0b2b16dc789" "-----BEGIN CERTIFICATE-----\nMIIDHDCCAgSgAwIBAgIICE1Jj+j4yf8wDQYJKoZIhvcNAQEFBQAwMTEvMC0GA1UE\nAxMmc2VjdXJldG9rZW4uc3lzdGVtLmdzZXJ2aWNlYWNjb3VudC5jb20wHhcNMTcw\nODIwMDA0NTI2WhcNMTcwODIzMDExNTI2WjAxMS8wLQYDVQQDEyZzZWN1cmV0b2tl\nbi5zeXN0ZW0uZ3NlcnZpY2VhY2NvdW50LmNvbTCCASIwDQYJKoZIhvcNAQEBBQAD\nggEPADCCAQoCggEBAI/PlhEYY9orCrm6np7s97QNPncHhTw0naBh2TiXa5/uq5fR\n9IncSbWle5rup2zuLSab8sPqpzsEFdw32nNFKm9MIcfE6+IxiJO33yWkaF4porPs\nQ3GQuTKt3xScLvhujkQGhj31erLNmXQsRD9wt1bUyM09NalhVbbfY/aqTMOyiqVq\nCfXsriBPLe9DnGncSZZ01VAd/+wYpErov8x4bRoa8/s/DmodsNnvxL4td+LDTMNz\nD3dM65ZSM6yXvo2ks36h9c9GghCk15bHLnidI+Sj0JI5GEcVWc98qS0BYE6WHqsf\nZUaUhIFhMTm/cqnsYG7GsZWvC7KLV5iP/ktZfKsCAwEAAaM4MDYwDAYDVR0TAQH/\nBAIwADAOBgNVHQ8BAf8EBAMCB4AwFgYDVR0lAQH/BAwwCgYIKwYBBQUHAwIwDQYJ\nKoZIhvcNAQEFBQADggEBAGCxjV13iiiCBRJHA0dwCmjiZ7g3Mr1HIEo0zac9f19T\n2k+7QavO5Z65pTxQJFs4Po9l7cQBkXXKlXlcXBY/o+oWgjWXHYV1rF1YH78soRDM\n8kKmjqJLL9QyHC0v6Lh07Miby3zcB8dHkeD3Z0Epwv/79kZuiQ15ziEgL+J2bHfD\nPJaRTMOy3oLHS8CSTa9B3Naahs2knJ1j3ZWZhb2vkRW985wBE9WPldihF7TI5Bnx\nFksm8g+WhJf2eFvMZ1Gv4MVKLd1TUpqaYbKy7W9Si73ZU+MN2m+hr8tZZPna63tB\n9K8lAr/wvORGO/bvyzmGatZfKO+YewLJp8eli/ZiFN0=\n-----END CERTIFICATE-----\n",
   "133ea9233b2d2699010c49da0642be6f8c946884" "-----BEGIN CERTIFICATE-----\nMIIDHDCCAgSgAwIBAgIIAlDfBLN1fs0wDQYJKoZIhvcNAQEFBQAwMTEvMC0GA1UE\nAxMmc2VjdXJldG9rZW4uc3lzdGVtLmdzZXJ2aWNlYWNjb3VudC5jb20wHhcNMTcw\nODIyMDA0NTI2WhcNMTcwODI1MDExNTI2WjAxMS8wLQYDVQQDEyZzZWN1cmV0b2tl\nbi5zeXN0ZW0uZ3NlcnZpY2VhY2NvdW50LmNvbTCCASIwDQYJKoZIhvcNAQEBBQAD\nggEPADCCAQoCggEBAKsOsphy7ZoqRqxgjQnJgYLpLAKNr4axnjlAM5YnDWLq3xwX\noyna8PMHuf3uqIVYUiZZCYkn4ynjxUyTlLm+5z6qstnJNaQqK8bSvDw9Fo7nFWcu\nNpJzVyq3EVtnHKUCkrDxAt2+iCax019r+j37h8y55OQjq1WLfavYkKmp27Kb0zpo\nPdYL/kmsJqb4TnsygFpSNjCCV/RrDIKz8DwhtuAuE7laqjocrU0jwMOhb/pTQM5g\np4LC23kIfJ9W/ALQpyZfbdjA7LmEngR4dMV690NmxkYqjaqf/us/qm/QMdbQdS0a\ntLBwYplCLjOuMWIozwbm2Kh3oY3kO74nxEEr+bcCAwEAAaM4MDYwDAYDVR0TAQH/\nBAIwADAOBgNVHQ8BAf8EBAMCB4AwFgYDVR0lAQH/BAwwCgYIKwYBBQUHAwIwDQYJ\nKoZIhvcNAQEFBQADggEBAC2lQP7pjia0ThCcPfQceTjXSSEiScF9ajLpfyFt/om3\nMvIFpf3I7XngORvmOSRQ8tApJ0qF3WYa5+rjyJatpvLj9OdBILegNPKw1tCsYI5X\n9GOtuoQpMTmTsEDYn/iXPHsh2Izk8Yh88bQQxDY5iNNQN0Yxu7e4ydyf2cmT8c3f\nQoc++snYsLKz0pw4K91GNGbR65fb2Q+pO/GuPiS8PqlGjmPLMZWHAEx0LUS2H7D5\nGNefNis7+XrS5Qq4HpiG9bcdTTOaUR2KhyNskjbeq1cOAU+CuuyCE3buAq7Zlktg\nTF7Zr+yWWV3uujYwgnCg3bUzyHWwuBBR9iKeHYSPnu0=\n-----END CERTIFICATE-----\n"})


(def cert-endpoint "https://www.googleapis.com/robot/v1/metadata/x509/securetoken@system.gserviceaccount.com")

(defn refresh-certs
  []
  (let [resp (client/get cert-endpoint)]
    (if (= 200 (:status resp))
      (json/decode (:body resp)))))


src/mast_session/handler.clj
(ns mast-session.handler
  (:require [mast-session.core :as sesh]
            [compojure.core :refer :all]
            [compojure.route :as route]
            [ring.middleware.defaults :refer [wrap-defaults site-defaults]]
            [ring.middleware.json :refer [wrap-json-response ]]
            [ring.util.response :refer [response]]
            [cheshire.generate :refer [add-encoder]]
            [clj-time.coerce :as convert]))

(add-encoder org.joda.time.DateTime
  (fn [dt jsonGenerator]
    (.writeString jsonGenerator (str (convert/to-epoch dt)))))

(defroutes json-routes
  (GET "/auth/:token" [token] (response (sesh/is-authorized? token)))
  (GET "/admin/sessions" [] (response @sesh/session)))

(defroutes app-routes
  (GET "/" [] "hello world")
  (route/not-found "nope"))

(defroutes all-routes
  (wrap-json-response json-routes {:pretty true})
  (wrap-defaults app-routes site-defaults))

(def app all-routes)


src/mast_session/session.clj
(ns mast-session.session
  (:require [clj-time.core :as t]))

(defonce mast-session (atom {}))

(def recent-threshold (t/minutes 3))

(defn recent?
  [t1 t2 threshold]
  (or (t/within? (t/interval t1 (t/plus t1 threshold)) t2)
    (t/within? (t/interval t2 (t/plus t2 threshold)) t1)))

(defn can-add-user?
  [user]
  (recent? (:auth-time user) (t/now) recent-threshold))

(defn mark-user-access
  [user]
  (assoc user :last-access (t/now)))

;; todo what should this return when you can't add the user?
(defn add-user
  [user session]
  (if (can-add-user? user)
    (swap! session #(assoc % (:user-id user) (mark-user-access user)))
    session))

(defn update-user-time
  [user session]
  (if-let [user-session (get @session (:user-id user))]
    (swap! session #(update % (:user-id user) (mark-user-access user-session))))
  session)

(defn has-session?
  [user session]
  (if-let [user-session (get @session (:user-id user))]
    (and (recent? (t/now) (:issued-at user) recent-threshold)
      (recent? (t/now) (:last-access user-session) recent-threshold))
    false))


src/mast_session/token.clj
(ns mast-session.token
  (:require [buddy.sign.jwt :as jwt]
            [buddy.sign.jws :as jws]
            [buddy.core.keys :as ks]
            [clj-time.core :as t]
            [clj-time.coerce :as convert]
            [mast-session.certs :as certs]))

(defn decode-google-payload
  [token cert opts]
  (try
    (jwt/unsign token cert opts)
    (catch Exception e nil)))

(def certs (map ks/str->public-key (vals certs/google-certs)))


(def default-token-options
  {:alg :rs256})
;; this won't need to map over all certs once we can see custom keys
;; in the header. See
;; https://github.com/funcool/buddy-sign/pull/52.
(defn decode-object
  ([token]
   (decode-object token default-token-options))
  ([token options]
   (->> certs
     (map #(decode-google-payload token % options))
     (filter some?)
     first)))

;; [:firebase :email :aud :sub :iss :exp :email_verified :auth_time :user_id :iat]


;; {:firebase
;;  {:identities {:email ["dsutton@breezeehr.com"]},
;;   :sign_in_provider "password"},
;;  :email "dsutton@breezeehr.com",
;;  :aud "breezeehr.com:breeze-ehr",
;;  :sub "Spgusa58rXSxAcfL0dab5yBP3Iz2",
;;  :iss "https://securetoken.google.com/breezeehr.com:breeze-ehr",
;;  :exp 1503090344,
;;  :email_verified false,
;;  :auth_time 1503086744,
;;  :user_id "Spgusa58rXSxAcfL0dab5yBP3Iz2",
;;  :iat 1503086744}

(defn firebase->user
  "Normalize timestamps"
  [firebase]
  (letfn [(epoch->joda [m v] (update m v convert/from-epoch))]
    (let [bad-names    (-> firebase
                         (epoch->joda :exp)
                         (epoch->joda :auth_time)
                         (epoch->joda :iat))
          renamed-keys {:iat            :issued-at
                        :exp            :expires
                        :iss            :issuer
                        :sub            :subject
                        :aud            :audience
                        :auth_time      :auth-time
                        :user_id        :user-id
                        :email_verified :email-verified}]
      (merge (select-keys bad-names [:identities :email])
        (reduce (fn [m [old-name new-name]]
                  (assoc m new-name (get bad-names old-name)))
          {}
          renamed-keys)))))

(defn token->user
  [token]
  (if-let [firebase (decode-object token)]
    (firebase->user firebase)
    nil))
